### PR TITLE
update and add bliss<->txt conversion jobs

### PR DESCRIPTION
**Pull request by JackTemaki**
- segment_file and gzip option for CorpusToTxtJob
 - added matching CorpusReplaceOrthFromTxtJob

### CODE REVIEW

File: corpus/convert.py [hunk 1]
```diff
@@ -79,11 +81,22 @@ def run(self):
...
+        :param Path corpus_path: Bliss corpus
+        :param Path segment_file: segment file
+        :param bool gzip: gzip the output text file
+        """
         self.set_vis_name("Extract TXT from Corpus")
 
         self.corpus_path = corpus_path
-        self.txt_path = self.output_path("corpus.txt")
+        self.gzip = gzip
+        self.segment_file = segment_file
```
Comment (curufinwe) [inline corpus/convert.py:112]: newline between member vars and outputs

File: corpus/convert.py [hunk 2]
```diff
@@ -93,6 +106,59 @@ def run(self):
         c = corpus.Corpus()
         c.load(corpus_path)
 
-        with open(self.txt_path.get_path(), "wt") as out:
-            for segment in c.segments():
+        if self.segment_file:
+            with open(tk.uncached_path(self.segment_file), "rt") as f:
+                lines = [l for l in f.readlines() if len(l.strip()) > 0]
```
Comment (curufinwe) [inline corpus/convert.py:125]: `lines = set(l for l in ...)` directly create the set from the iterator, no need for an intermediate list
also: why not strip the line that is added to the set?

File: corpus/convert.py [hunk 3]
```diff
@@ -79,11 +81,22 @@ def run(self):
...
+        :param Path segment_file: segment file
+        :param bool gzip: gzip the output text file
+        """
         self.set_vis_name("Extract TXT from Corpus")
 
         self.corpus_path = corpus_path
-        self.txt_path = self.output_path("corpus.txt")
+        self.gzip = gzip
+        self.segment_file = segment_file
+        self.out_txt = self.output_path("corpus.%s" % (".txt.gz" if gzip else ".txt"))
```
Comment (curufinwe) [inline corpus/convert.py:113]: creates two dots in the filename, better:
`self.out_txt = self.output_path("corpus.txt" + ".gz" if gzip else ""))`

File: corpus/convert.py [hunk 4]
```diff
@@ -93,6 +106,59 @@ def run(self):
...
+        self.out_corpus = self.output_path("corpus.xml.gz")
+
+    def tasks(self):
+        yield Task('run', mini_task=True)
+
+    def run(self):
+        c = corpus.Corpus()
+        c.load(tk.uncached_path(self.corpus_path))
+
+        if tk.uncached_path(self.text_file).endswith(".gz"):
```
Comment (curufinwe) [inline corpus/convert.py:165]: Use `uopen` https://github.com/rwth-i6/i6_core/blob/main/util.py#L144
Also: use `with` statement

File: corpus/convert.py [hunk 5]
```diff
@@ -93,6 +106,59 @@ def run(self):
...
+            text = gzip.open(tk.uncached_path(self.text_file), "r")
+        else:
+            text = open(tk.uncached_path(self.text_file), "rt")
+
+        for recording in c.all_recordings():
```
Comment (curufinwe) [inline corpus/convert.py:170]: You can directly iterate over all segments and then do something like:
```suggestion
        for segment, line in itertools.zip_longest(c.segments(), f)
           assert segment is not None
           assert line is not None
```

File: corpus/convert.py [hunk 6]
```diff
@@ -93,6 +106,59 @@ def run(self):
...
+
+        out.close()
+
+
+class CorpusReplaceOrthFromTxtJob(Job):
+    """
+    Merge raw text back into a bliss corpus
+    """
+
+    def __init__(self, corpus, text_file):
```
Comment (curufinwe) [inline corpus/convert.py:147]: This Job should work in tandem with the CorpusToTxtJob, i.e. support a list of segments to be passed into it as it relies on the line-number to match transcriptions to segments.

**Review by curufinwe — CHANGES_REQUESTED** [review]

File: corpus/convert.py [hunk 7]
```diff
@@ -93,6 +106,59 @@ def run(self):
...
+            out = open(self.out_txt.get_path(), "wt")
+
+        for segment in c.segments():
+            if not segments_whitelist or segment.fullname in segments_whitelist:
                 out.write(segment.orth + "\n")
+
+        out.close()
+
+
+class CorpusReplaceOrthFromTxtJob(Job):
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply corpus/convert.py:142]: already is?

File: corpus/convert.py (hunk 6 above)
Comment (JackTemaki) [reply corpus/convert.py:147]: I just copied the job from an old setup, but yes I could add that functionality.

File: corpus/convert.py (hunk 4 above)
Comment (JackTemaki) [reply corpus/convert.py:165]: I wasn't aware of that util function, so yes that would be much nicer.

File: corpus/convert.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (curufinwe) [reply corpus/convert.py:142]: sorry, looks like I only looked at one of the commits in the PR.

File: corpus/convert.py (hunk 6 above)
Comment (curufinwe) [reply corpus/convert.py:147]: :+1:

**Comment by JackTemaki** [comment]
The problem is I do not have a pipeline to test this now, I am not sure if we should merge it just like this.

File: corpus/convert.py [hunk 8]
```diff
@@ -93,6 +124,61 @@ def run(self):
...
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        c = corpus.Corpus()
+        c.load(tk.uncached_path(self.corpus_path))
+
+        if self.segment_file:
+            with open(tk.uncached_path(self.segment_file), "rt") as f:
+                segments_whitelist = set([l.strip() for l in f.readlines() if len(l.strip()) > 0])
```
Comment (curufinwe) [inline corpus/convert.py:172]: you are still creating an intermediate list, you can create the set from an iterator, no need for the list

File: corpus/convert.py [hunk 9]
```diff
@@ -93,6 +124,61 @@ def run(self):
...
 
-        with open(self.txt_path.get_path(), "wt") as out:
-            for segment in c.segments():
+        if self.segment_file:
+            with open(tk.uncached_path(self.segment_file), "rt") as f:
+                segments_whitelist = set([l.strip() for l in f.readlines() if len(l.strip()) > 0])
+        else:
+            segments_whitelist = None
+
+        if self.gzip:
```
Comment (curufinwe) [inline corpus/convert.py:133]: Actually you can also use uopen here, ".gz" is appended to path.

**Review by curufinwe — CHANGES_REQUESTED** [review]
Please don't force push, just add a new commit. For the final merge I will squash them into a single commit. If you force push I don't see the diff w.r.t. the previous commit.

**Review by curufinwe — APPROVED** [review]

**Pull request merged**